src/lib.rs

```
@@ -1,6 +1,349 @@
+static MMV_DIR_SUFFIX: &'static str = "mmv";
+
+type Endian = LittleEndian;
+
+#[cfg(unix)]
+fn get_process_id() -> i32 {
+    nix::unistd::getpid()
+}
+
+#[cfg(windows)]
+fn get_process_id() -> i32 {
+    unsafe { kernel32::GetCurrentProcessId() as i32 }
+}
+
+fn get_pcp_root() -> PathBuf {
+    match env::var_os("PCP_DIR") {
+        Some(val) => PathBuf::from(val),
+        None => PathBuf::from(MAIN_SEPARATOR.to_string())
+    }
+}
+
+fn init_pcp_conf(pcp_root: &PathBuf) -> io::Result<()> {
+    /* attempt to load variables from pcp_root/etc/pcp.conf into environment.
+       if pcp_root/etc/pcp.conf is not a file, can't be read, or parsing it
+       fails, we *don't* return the error */
+    let mut default_conf = pcp_root.clone();
+    default_conf.push("etc");
+    default_conf.push("pcp.conf");
+    parse_pcp_conf(default_conf).ok();
+
+    /* attempt to load variables from pcp_root/$PCP_CONF into environment.
+       if pcp_root/$PCP_CONF is not a file, can't be read, or parsing it
+       fails, we *do* return the error */
+    let mut pcp_conf = pcp_root.clone();
+    pcp_conf.push(env::var_os("PCP_CONF").unwrap_or(OsString::new()));
+    parse_pcp_conf(pcp_conf)
+}
+
+fn parse_pcp_conf(conf_path: PathBuf) -> io::Result<()> {
+    let pcp_conf = File::open(conf_path)?;
+    let buf_reader = BufReader::new(pcp_conf);
+
+    /* According to man 5 pcp.conf, syntax rules for pcp.conf are
+        1. general syntax is PCP_VARIABLE_NAME=value to end of line 
+        2. blank lines and lines begining with # are ignored
+        3. variable names that aren't prefixed with PCP_ are silently ignored
+        4. there should be no space between the variable name and the literal =
+        5. values may contain spaces and should not be quoted
+    */
+    lazy_static! {  
+        static ref RE: Regex =
+            Regex::new("^(PCP_[[:alnum:]_]+)=([^\"\'].*[^\"\']$)").unwrap();
+    }
+
+    for line in buf_reader.lines() {
```

Is `pcp.conf` guaranteed to be UTF-8?  If not, you need to read bytes instead of string `lines()`.

---

Since `parse_pcp_conf` returns from `&line?` if it reads invalid UTF-8, and all `io::Result<()>`s returned from `init_pcp_conf` are `.ok()`ed in the main code, is using `lines()` a problem?

---

If `lines()` reads invalid UTF-8, you'll get an `InvalidData` error.  But if `pcp.conf` is allowed to contain non-UTF-8 values, then this parser should allow that without error.

---

pcp.conf only ever contains ASCII content.

---

I mean "allow that" by still using those non-UTF-8 contents, not just `ok()`-squashing the error.

---

@natoscott Couldn't it contain arbitrary paths?

---

@cuviper yes it can, good point.

---

Okay, then I'll switch over to bytes.